### ISSUE TITLE
HTTP 500 after client connection

### ISSUE THREAD

**Issue by jfduque** [miguelgrinberg/python-engineio#118]
Got this exception after testing AsyncClient & AsyncServer. I'm using python-socketio, but the exception is raised in python-engineio.

Server side:
```python3
import socketio

sio = socketio.AsyncServer()
app = socketio.ASGIApp(sio)

@sio.event
def connect(sid, environ):
    print('connect ', sid)

@sio.event
def disconnect(sid):
    print('disconnect ', sid)```

```

Client side:
```python3
import asyncio
import socketio

sio = socketio.AsyncClient()

@sio.event
def connect():
    print("I'm connected!")

async def conectar():
    await sio.connect('http://localhost:8000')

loop = asyncio.get_event_loop()
loop.run_until_complete(conectar())
```

Server is deployed successfully using Gunicorn:
```bash
user@localhost:~$ venv/bin/gunicorn server:app -k uvicorn.workers.UvicornWorker -w 1
Starting gunicorn 19.9.0
[2019-06-13 16:49:10 +0000] [10109] [INFO] Listening at: http://127.0.0.1:8000 (10109)
[2019-06-13 16:49:10 +0000] [10109] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2019-06-13 16:49:10 +0000] [10112] [INFO] Booting worker with pid: 10112
[2019-06-13 16:49:10 +0000] [10112] [INFO] Started server process [10112]
[2019-06-13 16:49:10 +0000] [10112] [INFO] Waiting for application startup.
```
Running client throws ConnectionError exception:
```python3
socketio.exceptions.ConnectionError: Unexpected status code 500 in server response
```

And got the following exception on the server side
```python3
Traceback (most recent call last):
  File "/var/www/venv/lib/python3.6/site-packages/uvicorn/protocols/http/httptools_impl.py", line 368, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/var/www/venv/lib/python3.6/site-packages/engineio/async_drivers/asgi.py", line 46, in __call__
    await self.engineio_server.handle_request(scope, receive, send)
  File "/var/www/venv/lib/python3.6/site-packages/socketio/asyncio_server.py", line 313, in handle_request
    return await self.eio.handle_request(*args, **kwargs)
  File "/var/www/venv/lib/python3.6/site-packages/engineio/asyncio_server.py", line 181, in handle_request
    environ = translate_request(*args, **kwargs)
TypeError: translate_request() takes 1 positional argument but 3 were given

```
https://github.com/miguelgrinberg/python-engineio/blob/3457e4c8e189292a1d4b4139da618190de698258/engineio/asyncio_server.py#L181

Using `socketio.Server()` instead of `socketio.AsyncServer()` works fine.

My virtual environment contents:

- Click==7.0
- gunicorn==19.9.0
- h11==0.8.1
- httptools==0.0.13
- pkg-resources==0.0.0
- python-engineio==3.8.1
- python-socketio==4.1.0
- six==1.12.0
- uvicorn==0.7.2
- uvloop==0.12.2
- websockets==7.0

What will I be doing wrong? Thanks in advance Mr. Grinberg

**Comment by miguelgrinberg** [comment]
@jfduque This server supports the ASGI3 specification. In this version the ASGI application callable receives three arguments. The error suggests the web server (i.e. gunicorn) is using an older version of ASGI which sends one argument first, and a bit later the other two.

Try using uvicorn directly, that's what I use and works well for me.

**Comment by jfduque** [comment]
> @jfduque This server supports the ASGI3 specification. In this version the ASGI application callable receives three arguments. The error suggests the web server (i.e. gunicorn) is using an older version of ASGI which sends one argument first, and a bit later the other two.
> 
> Try using uvicorn directly, that's what I use and works well for me.

Hi Miguel, thanks for the quick response.

OK, I've tried to run the server using uvicorn:

```python3
import uvicorn
from application import app

uvicorn.run(app, host='0.0.0.0', port=8000, log_level="debug")

```
The server starts and is listening on port 8000. The server log is as follows:
```
INFO: Started server process [12497]
INFO: Waiting for application startup.
DEBUG: None - ASGI [1] Started
DEBUG: None - ASGI [1] Sent {'type': 'lifespan.startup'}
DEBUG: None - ASGI [1] Received {'type': 'lifespan.startup.complete'}
DEBUG: None - ASGI [1] Completed
INFO: Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
```

But when I start the client, another exception is raised on the server:

```
DEBUG: ('10.16.60.76', 31404) - Connected
DEBUG: ('10.16.60.76', 31404) - ASGI [2] Started
DEBUG: ('10.16.60.76', 31404) - ASGI [2] Raised exception
ERROR: Exception in ASGI application
Traceback (most recent call last):
  File "/var/www/venv/lib/python3.6/site-packages/uvicorn/protocols/http/httptools_impl.py", line 368, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/var/www/venv/lib/python3.6/site-packages/uvicorn/middleware/message_logger.py", line 58, in __call__
    raise exc from None
  File "/var/www/venv/lib/python3.6/site-packages/uvicorn/middleware/message_logger.py", line 54, in __call__
    await self.app(scope, inner_receive, inner_send)
  File "/var/www/venv/lib/python3.6/site-packages/engineio/async_drivers/asgi.py", line 46, in __call__
    await self.engineio_server.handle_request(scope, receive, send)
  File "/var/www/venv/lib/python3.6/site-packages/socketio/asyncio_server.py", line 313, in handle_request
    return await self.eio.handle_request(*args, **kwargs)
  File "/var/www/venv/lib/python3.6/site-packages/engineio/asyncio_server.py", line 181, in handle_request
    environ = translate_request(*args, **kwargs)
TypeError: translate_request() takes 1 positional argument but 3 were given
INFO: ('10.16.60.76', 31404) - "GET /socket.io/ HTTP/1.1" 500
```
I get the same result if I start the server from command line., like this:

`venv/bin/uvicorn --host 0.0.0.0 --port 8000 --log-level debug application:app`

**UPDATE:** 
I was checking your examples, when I came across the following line:

https://github.com/miguelgrinberg/python-socketio/blob/61034bb92200e0ed48cef307a6f9e982a802d9cf/examples/server/asgi/app.py#L8

I've replaced my `sio = socketio.AsyncServer()` for `sio = socketio.AsyncServer(async_mode='asgi')` and it's working now:

```
INFO: Started server process [12530]
INFO: Waiting for application startup.
INFO: Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
connect  13b45a445a9d44f3b5e3fe6dc405c500
INFO: ('10.16.60.76', 31794) - "GET /socket.io/ HTTP/1.1" 200
INFO: ('10.16.60.76', 31796) - "WebSocket /socket.io/" [accepted]
disconnect  13b45a445a9d44f3b5e3fe6dc405c500
```

**Comment by miguelgrinberg** [comment]
Hmm. Okay, I'll check the async mode autodetection, in general this is determined automatically but clearly that is not the case for ASGI.

**Comment by jfduque** [comment]
Fixed successfully in python-engineio 3.8.2.

As always, it is a delight to use your libraries. Thank you.
